machine.UART: Fix the double timeout with uart.read(), uart.write() and uart.readinto().

robert-hh opened the pull request:
### Summary

When uart.read(), uart.write() and uart_readinto() timed out because less data was transferred than expected, but some, the delay was uart.timeout + uart.timeout.char instead of just uart,timeout_char().

This PR changes the behavior into the expected one. It consists of 3 commits. The first changes uart.read() and uart.write(). The impact of that change is limited to the machine.UART class. The second commit changes stream.readinto() and may affect all uses of that function.
The third commit changes the ESP32's  machine_uart.c to deal properly with the timeout_char argument for uart.read().

uart.readline() is unchanged.

This PR addresses issue #17611.

### Testing

Tested with STM32(Pyboard V1.1 and PYBD_SF6), RP2 Pico, MIMXRT(Teensy 4.1), SAMD51, ESP32, NRF (nrf52840) and a CC3200 board using strings of varying length. Typical results:

```
RP2

UART(0, baudrate=38400, bits=8, parity=None, stop=1, tx=0, rx=1, txbuf=256, rxbuf=256, timeout=200, timeout_char=50, invert=None, irq=0)

uart.read() msg=""  timeout=201 ms
uart.read() msg="123"  timeout=52 ms
uart.read() msg="12345"  timeout=1 ms

uart.readline() msg=""  timeout=201 ms
uart.readline() msg="12345"  timeout=202 ms
uart.readline() msg="12345\n"  timeout=2 ms

uart.readinto(bytearray(5)) msg="" timeout=201 ms
uart.readinto(bytearray(5)) msg="123" timeout=52 ms
uart.readinto(bytearray(5)) msg="12345" timeout=1 ms

CC3200

UART(1, pins=(Pin("GP16"), Pin("GP17")))

uart.read() msg=""  timeout=36 ms
uart.read() msg="123"  timeout=38 ms (from 74 ms without the change)
uart.read() msg="12345"  timeout=5 ms

uart.readline() msg=""  timeout=35 ms
uart.readline() msg="12345"  timeout=41 ms
uart.readline() msg="12345\n"  timeout=6 ms

uart.readinto(bytearray(5)) msg="" timeout=35 ms
uart.readinto(bytearray(5)) msg="123" timeout=39 ms
uart.readinto(bytearray(5)) msg="12345" timeout=5 ms
```

### Trade-offs and Alternatives

If the impact of the changes to a stream module seems unsafe, the second commit may be omitted, causing the behavior of uart.readinto() unchanged.

robert-hh committed:
extmod/machine_uart: Return from read()/write() at the first timeout.

Do not try to read/write again after timeout happened once.

Fixes issue #17611.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
py/stream: Add a stream.readinto1() method for machine.UART.

Avoiding the double timeout when used with the UART class.
`stream.readinto1()` returns after the first timeout.

Fixes issue #17611.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
esp32/machine_uart: Handle properly the timeout_char argument.

Before, it was ignored.  Tested with ESP32, ESP32S3, ESP32C6.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented on py/stream.c:
```diff
@@ -303,7 +303,7 @@ static mp_obj_t stream_readinto(size_t n_args, const mp_obj_t *args) {
 
     int error;
-    mp_uint_t out_sz = mp_stream_read_exactly(args[0], bufinfo.buf, len, &error);
+    mp_uint_t out_sz = mp_stream_rw(args[0], bufinfo.buf, len, &error, MP_STREAM_RW_READ | MP_STREAM_RW_ONCE);
```
Sure. I can implement that.

robert-hh replied:
Can the `mp_stream_readinto1_obj` be simplified? like e.g. not accepting the second argument?
And should it have a configuration switch for boards with low memory, like `MICROPY_PY_STREAM_READINTO1`?

robert-hh replied:
PR updated with the change, which is just duplicating the `mp_stream_readinto_obj` with a single changed line. Size increase about 100 bytes among the ports.

robert-hh replied:
Done.  The size increase is now about 50 bytes.

robert-hh replied:
One behavior that is changed and may cause some irritation for users:
If `timeout_char` is not set, it is 0. Before it was then set to the duration of a byte. But that did not matter because it was effectively the same as `timeout`. Now the values matter, which was the intention of this change. 
Now I'm wondering if the default for `timeout_char` should be set to `timeout` to keep consistency with the previous behavior. However, that needs to change every port's `machine_uart` implementation and the docs.

dpgeorge replied:
> If `timeout_char` is not set, it is 0

I don't follow.  On eg the rp2 port, `timeout_char` is always at least as long as a single byte, even if the user sets it to 0.

I think the only behaviour that changes with the UART is that, after reading some bytes and timing out, it doesn't try to read more bytes (and hence wait another `timeout` time).

dpgeorge commented on py/stream.c:
```diff
@@ -303,7 +303,12 @@ static mp_obj_t stream_readinto(size_t n_args, const mp_obj_t *args) {
     int error;
-    mp_uint_t out_sz = mp_stream_read_exactly(args[0], bufinfo.buf, len, &error);
+    mp_uint_t out_sz;
+    if (flags & MP_STREAM_RW_ONCE) {
```
This `if` can be simplified to a single line:
```
mp_stream_rw(args[0], bufinfo.buf, len, &error, flags);
```
and then pass in `MP_STREAM_RW_READ` as flags from `stream_readinto()`, and `MP_STREAM_RW_READ | MP_STREAM_RW_ONCE` from `stream_readinto1()`.

robert-hh commented on extmod/machine_uart.c:
```diff
@@ -140,10 +140,10 @@ static const mp_rom_map_elem_t machine_uart_locals_dict_table[] = {
-    { MP_ROM_QSTR(MP_QSTR_readinto), MP_ROM_PTR(&mp_stream_readinto_obj) },
-    { MP_ROM_QSTR(MP_QSTR_write), MP_ROM_PTR(&mp_stream_write_obj) },
+    { MP_ROM_QSTR(MP_QSTR_readinto), MP_ROM_PTR(&mp_stream_readinto1_obj) },
+    { MP_ROM_QSTR(MP_QSTR_write), MP_ROM_PTR(&mp_stream_write1_obj) },
```
Which additional arguments are expected in the context of machine_uart's write()?

dpgeorge replied:
> It it safe then to change write1 to accept the additional arguments?

Yes, it's safe, because it's an extension of the existing signature.

Ideally we would not support these extra arguments, which are a MicroPython specific extension (extension to CPython).  Maybe we can remove them in MicroPython v2.0.

But for now, they are needed.  Eg `umqtt.simple` uses the second argument, eg `self.sock.write(pkt, 2)`.

robert-hh replied:
Change implemented & tested with a RP2 board. Both timeout and additional write() arguments work as expected.

robert-hh committed:
py/stream: Support additional arguments for mp_stream_write1_obj.

As suggested by @dpgeorge adding a `stream_write_generic()` function.

Tested the different write timeouts with a RP2 using flow control.  Tested
support of the additional arguments of `uart.write()`.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge reviewed (approved):
Thanks for updating, this looks good now.

Tested on PYBV10, the `machine.UART` tests all pass.
